Multiple changes:

JustinKyleJames opened the pull request:
* Added ModifyTime for list object.
* Make ListObjects work as AWS does.
* Undo multiple url encodes which were causing signature failures.
* On upload, checked for odstream open errors and returned INTERNAL_SERVER_ERROR if open failed.
* Removed compiler warnings for uncompleted methods without return codes. Returned false with a TODO instead.
* Fixed HeadObject by reading data object size and modify time.
* Fixed large file upload failures by adding body_limit().
* Added config params for read/write buffer sizes.
* Added support for the range header on GetObject.

korydraughn commented:
Nice. Let us know when this is ready for review.

korydraughn commented:
When you find some time, try rebasing your PR on top of #20.

See the commit message for what #20 currently does/changes. Overall, it's mostly about removing unnecessary steps now that we have Boost 1.81.0 as an externals package (in core-dev). However, it does change where the application looks for the config file. All you need to do is adjust your volume mount to look like the following:
```
docker run ... -v /path/to/config.json:/irods_client_s3_api.json:ro ...`
```

Keep in mind that rebasing will require a rebuild of the docker image, but that should be faster now that boost doesn't need to be compiled from source.

JustinKyleJames commented:
> When you find some time, try rebasing your PR on top of #20.
> 
> See the commit message for what #20 currently does/changes. Overall, it's mostly about removing unnecessary steps now that we have Boost 1.81.0 as an externals package (in core-dev). However, it does change where the application looks for the config file. All you need to do is adjust your volume mount to look like the following:
> 
> ```
> docker run ... -v /path/to/config.json:/irods_client_s3_api.json:ro ...`
> ```
> 
> Keep in mind that rebasing will require a rebuild of the docker image, but that should be faster now that boost doesn't need to be compiled from source.

Do you want me to push it after rebasing and testing?

JustinKyleJames commented:
I've split this up into several commits.  I think it is ready for review.

trel commented:
nice - thanks.

trel commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,200 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+
+            // Get the data objects within the collection
+            query = fmt::format(
+                "select COLL_NAME,DATA_NAME,DATA_OWNER_NAME,DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'",
```
please add spaces after commas

JustinKyleJames replied:
done

trel commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,200 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+
+            // Now get data objects 
+            query = fmt::format(
+                "select COLL_NAME,DATA_NAME,DATA_OWNER_NAME,DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'"
```
please add spaces after commas

JustinKyleJames replied:
done

trel commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,200 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+
+        // look for objects with COLL_NAME like <prefix>% 
+        query = fmt::format(
+            "select COLL_NAME,DATA_NAME,DATA_OWNER_NAME,DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME like '{}%'",
```
please add spaces after commas

JustinKyleJames replied:
done

trel commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,200 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+        for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
+            found_objects = true;
+            ptree object;
+            std::string key = (i[0].size() > base_length ? i[0].substr(base_length) : "") + "/" + i[1];
```
please add a comment to explain what is happening and why

JustinKyleJames replied:
I added a comment.  Please let me know if it is or is not sufficient.

trel commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,200 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+
+        // look for objects with COLL_NAME = <parent> and DATA_NAME like <object>% 
+        query = fmt::format(
+            "select COLL_NAME,DATA_NAME,DATA_OWNER_NAME,DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'"
```
please add spaces after commas

JustinKyleJames replied:
done

trel commented on src/s3/headobject.cpp:
```diff
@@ -49,8 +52,17 @@ boost::asio::awaitable<void> irods::s3::actions::handle_headobject(
+                auto last_write_time__time_point = irods::experimental::filesystem::client::last_write_time(*thing, path);
+                std::time_t last_write_time__time_t = std::chrono::system_clock::to_time_t(last_write_time__time_point);
+                std::string last_write_time__str = irods::s3::api::common_routines::convert_time_t_to_str(last_write_time__time_t, "%a, %d %b %Y %H:%M:%S %Z");
+                //response.set("Last-Modified", "Mon, 2 Jan 2007 15:04:05 GMT");
```
can remove

JustinKyleJames commented:
Also note that I can remove the commit about compiler warnings if that is preferable to TODO's.

trel commented on src/s3/getobject.cpp:
```diff
@@ -115,14 +114,13 @@ asio::awaitable<void> irods::s3::actions::handle_getobject(
                         // It would be nice if we could figure out something a bit more
                         // semantic than catching an exception
-                        std::cout << "Good error!" << std::endl;
+                        //std::cout << "Good error!" << std::endl;
```
can remove

JustinKyleJames replied:
done

trel commented on src/connection.cpp:
```diff
@@ -58,3 +60,22 @@ std::string irods::s3::get_resource()
+
+uint64_t irods::s3::get_read_buffer_size_in_bytes()
+{
+    //port = s3_server.value("port", 8080);
```
can remove

JustinKyleJames replied:
done

trel commented on src/s3/putobject.cpp:
```diff
@@ -77,23 +78,26 @@ asio::awaitable<void> irods::s3::actions::handle_putobject(
+        uint64_t read_buffer_size = irods::s3::get_read_buffer_size_in_bytes();
+        std::cout << "read buffer size = " << read_buffer_size << std::endl;
+        std::vector<char> buf_vector(read_buffer_size);
+        //char buf[4096];
```
can remove

trel reviewed (commented):
Looking very nice.

Do we have a logging library/framework in mind?    Seeing more `cout` than I expected.

trel commented:
> Also note that I can remove the commit about compiler warnings if that is preferable to TODO's.

@korydraughn you have an opinion on this one?

korydraughn commented:
Will review this weekend.

JustinKyleJames commented:
> Looking very nice.
> 
> Do we have a logging library/framework in mind? Seeing more `cout` than I expected.

Agree that we should add on.  We should discuss.

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                full_path.parent_path().string(),
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
```
`i` -> `row`

JustinKyleJames replied:
done for all

JustinKyleJames replied:
done

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                full_path.parent_path().string(),
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
```
Consider renaming `thing` to something that expresses what it is.

JustinKyleJames replied:
done

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                found_objects = true;
+                ptree object;
+                std::string key = (i[0].size() > base_length ? i[0].substr(base_length) : "");
+                if (key.rfind("/", 0) == 0) {
```
Does this do what you intended it to do?

What's the goal of this line?

JustinKyleJames replied:
I didn't write that part originally but I believe the intent is to remove a leading slash if it exists.

korydraughn replied:
Ok. That line looks suspicious to me. It's a reverse search starting at the beginning of the string?

JustinKyleJames replied:
Yes, it is.  Kind of odd.

JustinKyleJames replied:
I changed this to ` if key.starts_with("/")`

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+            // Get the data objects within the collection
+            query = fmt::format(
+                "select COLL_NAME, DATA_NAME, DATA_OWNER_NAME, DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'",
+                full_path.parent_path().string());
```
`.string()` -> `.c_str()`

JustinKyleJames replied:
Same as earlier.  This is a path object not a std::string.

korydraughn replied:
`full_path.parent_path().c_str()`

JustinKyleJames replied:
done

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                "select COLL_NAME, DATA_NAME, DATA_OWNER_NAME, DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'",
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
```
- `i` -> `row`
- Rename `thing`

JustinKyleJames replied:
done for both

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                object.put("Owner", i[2]);
+                object.put("Size", atoi(i[3].c_str()));
+                try {
+                    std::time_t modified_epoch_time = boost::lexical_cast<std::time_t>(i[4]);
```
Why use boost rather than the standard here?

`boost::lexical_cast` seems a little overkill given we have `std::sto*` functions in the standard library.

See https://en.cppreference.com/w/cpp/string/basic_string/stol. There are more options at the bottom of that page too.

JustinKyleJames replied:
I believe it is because the boost libraries will error out of they don't gobble up all of the input.  IMO the following should produce an exception.

```
#include <string>
#include <iostream>

int main()
{
    std::string s("124abc");
    
    try {
        int i = std::stoi(s);
        std::cout << i << std::endl;
    } catch (std::invalid_argument const& ex) {
        std::cout << "invalid_argument - " << ex.what() << std::endl;
    }

    return 0;
}
```

And the output:

```
124


...Program finished with exit code 0
Press ENTER to exit console.
```

korydraughn replied:
That's fair.

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                full_path.string(),
+                full_path.string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
```
See previous comments for other loops.

JustinKyleJames replied:
done

korydraughn commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+            query = fmt::format(
+                "select COLL_NAME, DATA_NAME, DATA_OWNER_NAME, DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'"
+                " and DATA_NAME like '{}%'",
+                full_path.parent_path().string(),
+                full_path.object_name().string());
```
`.c_str()`

JustinKyleJames replied:
Same as before.  This is a path object not a std::string.

JustinKyleJames replied:
done

korydraughn commented on src/s3/putobject.cpp:
```diff
@@ -110,4 +119,4 @@ asio::awaitable<void> irods::s3::actions::handle_putobject(
 }
-#pragma clang diagnostic pop
\ No newline at end of file
+#pragma clang diagnostic pop
```
Let's plan to get rid of the `#pragma`s. Please capture in an issue.

JustinKyleJames replied:
done

korydraughn reviewed (commented):
It's coming together.

JustinKyleJames commented on src/authentication.cpp:
```diff
@@ -96,9 +96,9 @@ namespace
+                        return std::pair<std::string, std::string>(a.key, a.value);
                     }
-                    return std::make_pair(uri_encode(a.key), std::string(""));
+                    return std::pair<std::string, std::string>(a.key, std::string(""));
```
done

JustinKyleJames commented on src/common_routines.hpp:
```diff
@@ -0,0 +1,16 @@
+#ifndef IRODS_S3_API_COMMON_ROUTINES_HPP
+#define IRODS_S3_API_COMMON_ROUTINES_HPP
+#include <boost/beast.hpp>
```
done

JustinKyleJames commented on src/common_routines.hpp:
```diff
@@ -0,0 +1,16 @@
+namespace irods::s3::api::common_routines {
+
+
+    static std::string convert_time_t_to_str(time_t t, const std::string format) {
```
done

JustinKyleJames commented on src/common_routines.hpp:
```diff
@@ -0,0 +1,16 @@
+namespace irods::s3::api::common_routines {
+
+
+    static std::string convert_time_t_to_str(time_t t, const std::string format) {
```
No.  Done.

JustinKyleJames commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
 
-    std::string filename_prefix = "%";
-
+    irods::experimental::filesystem::path the_prefix(""); 
```
done

JustinKyleJames commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
-        beast::http::response<beast::http::empty_body> response;
-        response.result(beast::http::status::not_found);
-        beast::http::write(socket, response);
+        if (full_path.object_name().string() == "") {
```
done

JustinKyleJames commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+            // Get exact collections underneath this collection 
+            query = fmt::format(
+                "select COLL_NAME where COLL_NAME like '{}/%' and COLL_NAME not like '{}/%/%'",
+                full_path.parent_path().string(),
+                full_path.parent_path().string());
```
I am unsure what the request is here.  The string() is required (I think) because full_path.parent_path() is a irods::experimental::filesystem::path object.

korydraughn replied:
Sorry.

`.string()` -> `.c_str()`

JustinKyleJames replied:
done

JustinKyleJames commented on src/connection.cpp:
```diff
@@ -8,6 +8,8 @@ namespace
     nlohmann::json g_config;
 
     std::optional<std::string> resource;
+    std::optional<uint64_t> read_buffer_size_in_bytes;
+    std::optional<uint64_t> write_buffer_size_in_bytes;
```
done

JustinKyleJames commented on src/connection.cpp:
```diff
@@ -58,3 +60,21 @@ std::string irods::s3::get_resource()
     return resource.value();
 }
+
+uint64_t irods::s3::get_read_buffer_size_in_bytes()
+{
+    if (!read_buffer_size_in_bytes.has_value()) {
+        read_buffer_size_in_bytes = g_config.value(
+                nlohmann::json::json_pointer{"/s3_server/read_buffer_size_in_bytes"}, 8192);
+    }
+    return read_buffer_size_in_bytes.value();
+}
+
+uint64_t irods::s3::get_write_buffer_size_in_bytes()
+{
+    if (!write_buffer_size_in_bytes.has_value()) {
+        write_buffer_size_in_bytes = g_config.value(
+                nlohmann::json::json_pointer{"/s3_server/write_buffer_size_in_bytes"}, 8192);
+    }
+    return write_buffer_size_in_bytes.value();
+}
```
done

JustinKyleJames commented on src/connection.hpp:
```diff
@@ -34,5 +34,9 @@ namespace irods::s3
 
     std::string get_resource();
+
+    uint64_t get_read_buffer_size_in_bytes();
+    uint64_t get_write_buffer_size_in_bytes();
```
done

alanking commented on src/common_routines.hpp:
```diff
@@ -0,0 +1,15 @@
+
+    inline std::string convert_time_t_to_str(time_t t, const std::string& format) {
+        // example 2009-10-12T17:50:30.000Z
+        char buffer[80];
+        strftime(buffer, 80, format.c_str(), localtime(&t));
+        return buffer;
```
Could we use libfmt here? I might be missing the context.

korydraughn replied:
Seems you may be able to do that. I'm sure we've use it in the iRODS server somewhere.

See https://fmt.dev/8.1.1/api.html#date-and-time-formatting.

Up to you guys.

alanking replied:
Shiny!

korydraughn replied:
You can get around the compile-time issues by using `fmt::runtime(...)`. For example:
```cpp
const auto s = fmt::format(fmt::runtime("some format string {}"), 42);
```

JustinKyleJames replied:
That is much better.  I'll make that change.

JustinKyleJames replied:
I made the change to use fmt::runtime() and also adding back in the common_routines.hpp.  Please take a look.

I force pushed over the last commit so that the last one now will be clean.

korydraughn replied:
The change looks good.

If @alanking agrees, feel free to mark this as resolved.

alanking replied:
Yep, looks much better. Resolve it!

alanking commented on src/configuration.cpp:
```diff
@@ -0,0 +1,53 @@
+
+void irods::s3::set_config(const nlohmann::json& _config)
+{
+    g_config = _config;
```
Would it be worth making this thread safe?

korydraughn replied:
Perhaps in the future. For now, we're choosing to keep the config static on program start to keep things simple and deterministic.

alanking commented on src/s3/listobjects.cpp:
```diff
@@ -32,18 +35,19 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
     using namespace boost::property_tree;
 
-    auto thing = irods::s3::get_connection();
+    auto rcComm_t_ptr = irods::s3::get_connection();
```
If this is a pointer, should we use `auto*` here?

korydraughn replied:
Hmm, is that really a pointer?

alanking replied:
Not sure. It says "_ptr" in the name and it is being dereferenced with an asterisk later, so it checked several boxes :)

JustinKyleJames replied:
It is a smart pointer (std::unique_ptr).  Will auto* work with smart pointers?

JustinKyleJames replied:
Just verified that auto* will not work with the smart pointer.

alanking commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                std::string modified_time_str = irods::s3::api::common_routines::convert_time_t_to_str(modified_epoch_time, "%Y-%m-%dT%H:%M:%S.000Z");
+                object.put("LastModified", modified_time_str);
+            } catch ( const boost::bad_lexical_cast& ) {
+                // do thing - don't add LastModified tag
```
do thing -> do nothing?

JustinKyleJames replied:
Yes, search/replace error I believe.

JustinKyleJames replied:
Done -  in multiple locations.

alanking commented on src/s3/listobjects.cpp:
```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
+                std::string modified_time_str = irods::s3::api::common_routines::convert_time_t_to_str(modified_epoch_time, "%Y-%m-%dT%H:%M:%S.000Z");
+                object.put("LastModified", modified_time_str);
+            } catch ( const boost::bad_lexical_cast& ) {
+                // do norcComm_t_ptr - don't add LastModified tag
```
do norcComm_t_ptr?

JustinKyleJames replied:
Ugh.  Bad search/replace.

JustinKyleJames replied:
Done

alanking reviewed (commented):
Seems good overall, the parts that I could understand :)

korydraughn commented on src/s3/headobject.cpp:
```diff
@@ -14,6 +14,8 @@ namespace asio = boost::asio;
 namespace beast = boost::beast;
 namespace fs = irods::experimental::filesystem;
 
+static const std::string date_format{"{:%a, %d %b %Y %H:%M:%S %Z}"};
```
So the date format changes depending on the S3 API call?

JustinKyleJames replied:
From testing with aws command line tool using the --debug flag, that is what I got.  The aws command line tool seems to like it and converts the dates correctly to its output format for each different API.

korydraughn commented:
Looks like all review comments have been resolved.

Are the changes passing tests? What's left for this PR?

trel commented:
Please squash as appropriate.

JustinKyleJames commented:
I just committed an addition to support the range header on GetObject.  Please check the latest commit.

korydraughn commented on src/s3/getobject.cpp:
```diff
@@ -98,8 +154,11 @@ asio::awaitable<void> irods::s3::actions::handle_getobject(
                 response.result(beast::http::status::ok);
-                d.read(buf_vector.data(), write_buffer_size);
+
+                std::size_t read_length = write_buffer_size < range_end+1 - offset ? write_buffer_size : range_end+1 - offset;
```
The ternary expression is a little difficult to read.

Consider the following:
- Using parens to group expressions
- Or changing it to use an if-ladder

JustinKyleJames replied:
I broke it into multiple lines and added comments.

korydraughn commented on src/s3/getobject.cpp:
```diff
@@ -112,14 +171,18 @@ asio::awaitable<void> irods::s3::actions::handle_getobject(
                 catch (boost::system::system_error& e) {
                     if (e.code() != beast::http::error::need_buffer) {
-                        std::cout << "Not a good error!" << std::endl;
+                        std::cout << "Not a good error!" << e.code() << std::endl;
```
I think `system_error` contains a message as well. Let's print that out too.

Also, let's change the `Not a good error!` to something that helps the admin better or consider removing it altogether.

JustinKyleJames replied:
> I think `system_error` contains a message as well. Let's print that out too.
> 
> Also, let's change the `Not a good error!` to something that helps the admin better or consider removing it altogether.

Yep, I didn't write that.  I'll change the message and add e.code().message().

JustinKyleJames replied:
I changed the message and added e.code().message() to the output.

korydraughn commented on src/s3/getobject.cpp:
```diff
@@ -112,14 +171,18 @@ asio::awaitable<void> irods::s3::actions::handle_getobject(
                     }
                 }
+
+                if (offset > range_end) {
+                    break;
+                }
```
Consider adding a comment explaining this.

JustinKyleJames replied:
Added comment.

korydraughn reviewed (commented):
The last commit looks fine.

I'm guessing the test for this will appear in the other PR?

JustinKyleJames commented:
> The last commit looks fine.
> 
> I'm guessing the test for this will appear in the other PR?

Yes it will be in the other PR.

JustinKyleJames commented:
@korydraughn - I updated the README and also changed the name of the buffer size configuration options in the last two commits.  The buffer configuration options commit can be merged with the original commit once given the go-ahead.

SwooshyCueb commented on Dockerfile:
```diff
@@ -24,7 +24,7 @@ RUN apt-get update && \
         help2man \
         libbz2-dev \
-        libcurl4-gnutls-dev \
+        libcurl4-openssl-dev \
```
Why is this change needed?

korydraughn replied:
For some reason, that change allows @JustinKyleJames to build the image AFAIK. I'm able to build without that change, so it's possible it could be something in his environment. 

@JustinKyleJames Please capture the build output you get into the issue you created. @SwooshyCueb may be able to see the problem.

JustinKyleJames replied:
I just added it to that issue.  Also will copy it here:

```
dpkg: dependency problems prevent configuration of irods-runtime:
 irods-runtime depends on libcurl4; however:
  Package libcurl4 is not installed.

dpkg: error processing package irods-runtime (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 irods-runtime
```

korydraughn commented on README.md:
```diff
@@ -39,6 +39,8 @@ Goal is to support the equivalent of:
 Multipart uploads are an important part of S3, however, they aren't simple to implement and have substantial performance
 and storage implications for a naive approach. As of now, this API does not attempt to provide that functionality.
 
+See [To disable multipart in clients](#disabling-multipart)
```
How about ...

`See [Disabling Multipart](#disabling-multipart) for details.`

JustinKyleJames replied:
Done.

korydraughn commented on README.md:
```diff
@@ -151,6 +153,12 @@ The following code block shows the structure of the configuration file and provi
+        // The size of the buffer when calling PutObject.
+        "put_object_buffer_size_in_bytes": 8192,
+
+        // The size of the buffer when calling GetObject 
```
Missing a period at the end of the sentence.

JustinKyleJames replied:
Well, it is technically not a proper sentence but I will add the period.

JustinKyleJames replied:
I added the period.  I always struggle when I am writing a sentence fragment but agree consistency is important.

korydraughn commented on README.md:
```diff
@@ -151,6 +153,12 @@ The following code block shows the structure of the configuration file and provi
+        "put_object_buffer_size_in_bytes": 8192,
+
+        // The size of the buffer when calling GetObject 
+        "get_buffer_size_in_bytes": 8192 
```
`get_buffer_size_in_bytes` -> `get_object_buffer_size_in_bytes`

JustinKyleJames replied:
Made change.

korydraughn reviewed (commented):
I think we'll be in a good spot for merging after this review.

Let's see what @SwooshyCueb thinks of the libcurl-dev build issue before merging. He may have some suggestions for you.

korydraughn commented:
Noting here that I was able to build this PR without any issues. The build only used a cached version of the base OS image for Ubuntu 20.04.

korydraughn commented:
@JustinKyleJames Looks like we're ready for squashing.

korydraughn reviewed (approved):
If you're happy with it and the tests pass, pound it up.

JustinKyleJames commented:
This had been pounded and is ready for one final review.

korydraughn commented:
Given all of the review comments are resolved. I think we're good to merge.
